File: P5 2024-2/Paradigmas de Programacao/09-06/Python/Carro.py
class Motor:
    def __init__(self, tipo, potencia):
        self.tipo = tipo
        self.potencia = potencia

    def ligar(self):
        print("O motor está ligado.")

    def desligar(self):
        print("O motor está desligado.")
        
class Pneu:
    def __init__(self, marca, tamanho):
        self.marca = marca
        self.tamanho = tamanho

    def inflar(self):
        print("O Pneu está inflado.")

    def desinflar(self):
        print("O Pneu está desinflado.")

class Carro:
    def __init__(self, marca, modelo):
        self.marca = marca
        self.modelo = modelo
        self.motor = Motor("Gasolina", 150)
        self.pneu = [Pneu("Pirelli", 18) for _ in range(4)]
 
    def ligar(self):
        self.motor.ligar()
        print("O carro está pronto para rodar.")

    def desligar(self):
        self.motor.desligar()
        print("O carro está desligado.")

    def trocar_pneu(self, indice, novo_pneu):
        pneu_removido = self.pneu[indice]
        self.pneu[indice] = novo_pneu
        print(f"Pneu {indice + 1} foi trocado")
        return pneu_removido
    
carro = Carro("Toyota", "Corolla")
carro.ligar()
novo_pneu = Pneu("Michelin", 17)
pneu_removido = carro.trocar_pneu(2, novo_pneu)
print("Pneu removido: ", pneu_removido.marca)
carro.desligar()

File: P5 2024-2/Paradigmas de Programacao/09-06/Golang/Carro.go
package main

import "fmt"

type Motor struct {
	tipo     string
	potencia int
	ligado   bool
}

func (m *Motor) ligar() {
	m.ligado = true
	fmt.Println("O Motor está ligado")
}

func (m *Motor) desligar() {
	m.ligado = false
	fmt.Println("O Motor está desligado")
}

type Roda struct {
	marca   string
	tamanho int
	inflado bool
}

func (r *Roda) inflar() {
	r.inflado = true
	fmt.Println("A Roda está inflada")
}

func (r *Roda) desinflar() {
	r.inflado = false
	fmt.Println("A Roda está desinflada")
}

type Carro struct {
	Motor
	Rodas  []Roda
	ligado bool
}

func (c *Carro) ligarCarro() int {
	if !c.Motor.ligado {
		fmt.Println("Não é possível ligar o carro, o motor está desligado")
		return 0
	}

	for _, r := range c.Rodas {
		if !r.inflado {
			fmt.Println("Não é possível ligar o carro, tem pelo menos uma roda desinflada")
			return 0
		}
	}

	c.ligado = true
	fmt.Println("O Carro está ligado")
	return 1
}
